### ISSUE TITLE
Support input-based loss-functions in tf.keras.Model

### ISSUE THREAD

**Issue by viktor765** [tensorflow/tensorflow#38353]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): TF 2.2.0rc2
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

In `tf.keras.Model`, there should be an API for supplying loss functions and metrics that are based on the input `x_input` and not just the output `y_pred`. 

Currently when supplying custom loss functions in `tf.keras.Model` or metrics they need to be on the form `loss(y_true, y_pred)`. The most import example is in `Model.compile()` which accepts a functional argument `loss` of this type but not one having access to `x_input`.

**Will this change the current api? How?**

Yes, but there are many possible implementations in which compatibility can be preserved. For example `Model.compile()` could accept an argument `loss_x` (supplying both this and `loss` in the same call throws an error). This should then accept a function as follows.

`def custom_loss_x(y_true, y_pred, x_input): ... `

or preferably

`def custom_loss_x(y_true, x_input): ... `

as y_pred can be retained by `y_pred = model(x_input)`.

In an analogous way custom metrics for models can be constructed and supplied.

**Who will benefit with this feature?**

Any model with a loss function or metrics that uses information available in the input layer but not in the output layer.

An example is when the input consists of a segmented image and the output is an image of the same shape. If different segments are to be weighted differently in the loss function this loss function will depend on `x_input`. 

**Any Other info.**

**Comment by jvishnuvardhan** [comment]
@viktor765 Can you please try `model.add_loss(y_true, y_pred, weights)` as in the following example. 

```
import tensorflow as tf
from tensorflow.keras.layers import Dense, Input
from tensorflow.keras.models import Model
from tensorflow.keras.losses import Loss
import numpy as np
 
inp = Input(shape=(1,), dtype=tf.float32)
targets = Input(shape=(1,), dtype=tf.float32)
w = Input(shape=(1,), dtype=tf.float32)
out = Dense(1)(inp)
m = Model(inputs=[inp, w, targets], outputs=out)
 
 
 
def weighted_loss(y_true, y_pred, weights):
      loss = tf.math.squared_difference(y_pred, y_true)
      w_loss = tf.multiply(weights, loss)
      return tf.reduce_mean(tf.reduce_sum(w_loss, axis=-1))
  
m.add_loss(weighted_loss(targets, out, inp))
 
m.compile(loss=None, optimizer='adam')
x = np.ones((512,1))
m.fit([x, x, x], epochs=10)
```

**Comment by viktor765** [comment]
@jvishnuvardhan That achieves precisely what I want and more, thank you. I was unaware of Model.add_loss but I think its versatility makes it the best way of defining loss functions.  

Perhaps it should be more present in the Tensorflow documentation? I believe that jlh's answer here explains its case quite well: https://stackoverflow.com/questions/50063613/what-is-the-purpose-of-the-add-loss-function-in-keras
